## Исследование надёжности заёмщиков

### Открываем файл с данными, изучаем 

In [1]:
import pandas as pd #импортируем Пандас
data = pd.read_csv('/datasets/data.csv') #читаем файл с данными

from pymystem3 import Mystem #импортируем Mystem - позднее будет нужен для лемматизации 
m = Mystem()

print(data.info()) #смотрим структуру таблицы
data.head(30) #смотрим первые записи

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB
None


children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   
20         1   -1311.604166         36              среднее             1   
21         1    -253.685166         33              среднее             1   
22         1   -1766.644138         24              среднее             1   
23         0    -272.981385         21               высшее             0   
24         1  338551.952911         57              среднее             1   
25         0  363548.489348         67              среднее             1   
26         0            NaN         41              среднее             1   
27         0    -529.191635         28               высшее             0   
28         1    -717.274324         26               высшее             0   
29         0            NaN         63              среднее             1   

            family_status  family_status_id gender  income_type  debt  \
0         женат / замужем                 0      F    сотрудник     0   
1         женат / замужем                 0      F    сотрудник     0   
2         женат / замужем                 0      M    сотрудник     0   
3         женат / замужем                 0      M    сотрудник     0   
4        гражданский брак                 1      F    пенсионер     0   
5        гражданский брак                 1      M    компаньон     0   
6         женат / замужем                 0      F    компаньон     0   
7         женат / замужем                 0      M    сотрудник     0   
8        гражданский брак                 1      F    сотрудник     0   
9         женат / замужем                 0      M    сотрудник     0   
10        женат / замужем                 0      M    компаньон     0   
11        женат / замужем                 0      F    сотрудник     0   
12       гражданский брак                 1      M    пенсионер     0   
13        женат / замужем                 0      F    сотрудник     0   
14       гражданский брак                 1      F    компаньон     1   
15        женат / замужем                 0      F    сотрудник     0   
16        женат / замужем                 0      F    сотрудник     0   
17       гражданский брак                 1      M    сотрудник     0   
18         вдовец / вдова                 2      F    пенсионер     0   
19              в разводе                 3      F    сотрудник     0   
20        женат / замужем                 0      M    сотрудник     0   
21       г

In [2]:
#смотрим, какие уникальные значения есть у столбцов, нужных для проекта: 
print('Количество детей\n',data['children'].unique()) #количество детей
print('\nСемейное положение\n',data['family_status'].unique()) #семейное положение
print('\nЗадолженность\n',data['debt'].unique()) #задолженность
print('\nЦель получения кредита\n',data['purpose'].unique()) #цель получения кредита 

Количество детей
 [ 1  0  3  2 -1  4 20  5]

Семейное положение
 ['женат / замужем' 'гражданский брак' 'вдовец / вдова' 'в разводе'
 'Не женат / не замужем']

Задолженность
 [0 1]

Цель получения кредита
 ['покупка жилья' 'приобретение автомобиля' 'дополнительное образование'
 'сыграть свадьбу' 'операции с жильем' 'образование'
 'на проведение свадьбы' 'покупка жилья для семьи' 'покупка недвижимости'
 'покупка коммерческой недвижимости' 'покупка жилой недвижимости'
 'строительство собственной недвижимости' 'недвижимость'
 'строительство недвижимости' 'на покупку подержанного автомобиля'
 'на покупку своего автомобиля' 'операции с коммерческой недвижимостью'
 'строительство жилой недвижимости' 'жилье'
 'операции со своей недвижимостью' 'автомобили' 'заняться образованием'
 'сделка с подержанным автомобилем' 'получение образования' 'автомобиль'
 'свадьба' 'получение дополнительного образования' 'покупка своего жилья'
 'операции с недвижимостью' 'получение высшего образования'
 'свой авто

### Вывод

Всего 21525 записей.

Из замеченных недочетов в файле с данными:

1. Столбец education - разный регистр (например: СРЕДНЕЕ, среднее);
2. Столбец days_employed - много отрицательных значений (скорее всего, при заполнении перепутали поля начальной и конечной дат), есть отсутствующие значения (NaN) (видимо, забыли спросить), тип данных - float64;
3. Столбец purpose - некоторые цели описаны по разному, но суть одна (на покупку своего автомобиля, сделка с автомобилем, на покупку автомобиля и пр.);
4. Столбец total_income	- есть отсутствующие значения (NaN) (видимо, забыли спросить), тип данных - float64;
5. Столбец dob_days (возраст в годах) содержит значения 0;
6. Столбец children - есть значения -1 и 20;
7. Столбец gender - помимо значений M и F есть запись со значением XNA.

NaN в days_employded и в total_income совпадают по строкам.

### Шаг 2. Предобработка данных

### Обработка пропусков

In [3]:
def FillnaTotalIncome(param):
    median_data_ti = data[data['income_type'] == param]['total_income'].median()
    data.loc[data['income_type']==param,'total_income'] = data.loc[data['income_type']==param,'total_income'].fillna(median_data_ti)
    return data
    
#заполняем пропуске в столбце total_income

#смотрим, сколько всего у нас пропусков в total_income
print('Всего пропусков в total_income\n',data[data['total_income'].isnull()].count()) 

#выясняем, как пропуски в total_income распределены по типам занятости
print('\nРаспределение пропусков по типам занятости\n',data['total_income'].isnull().groupby(data['income_type']).sum().sort_values()) 

#список типов занятости, у которых есть пропуски в total_income
NaN_income_type=['госслужащий','компаньон','пенсионер','предприниматель','сотрудник'] 

#пропуски заполним медианными значениями. 
#считаем медиану, а не среднее, потому что большой разброс между максимальным и минимальным значением дохода
#поскольку у разных типов занятости разные уровни дохода, рассчитаем медиану для каждого типа занятости
for q in NaN_income_type:
    data = FillnaTotalIncome(q)

#проверяем, остались ли пустые значения в total_income
print('\nПропущенных значений в total_income после заполнения:','\n',data[data['total_income'].isnull()].count()) 


Всего пропусков в total_income
 children            2174
days_employed          0
dob_years           2174
education           2174
education_id        2174
family_status       2174
family_status_id    2174
gender              2174
income_type         2174
debt                2174
total_income           0
purpose             2174
dtype: int64

Распределение пропусков по типам занятости
 income_type
безработный           0.0
в декрете             0.0
студент               0.0
предприниматель       1.0
госслужащий         147.0
пенсионер           413.0
компаньон           508.0
сотрудник          1105.0
Name: total_income, dtype: float64

Пропущенных значений в total_income после заполнения: 
 children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64


### Вывод

Заполнили пропуски в столбце total_income медианными значениями, рассчитанными для каждого типа занятости.

Еще можно было бы сделать следующее:
- заполнить пропуски в days_employed и перевести отрицательные значения в положительные;
- заменить в столбце dob_dyears (возраст в годах) значение 0 на медианное значение.

Но поскольку эти столбцы нам не нужны в рамках даного проекта, не будем этим заниматься.


### Замена типа данных

In [4]:
data['total_income'] = data['total_income'].astype('int') #total_income переводим из float64 в in64
data.info() #проверяем

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null int64
purpose             21525 non-null object
dtypes: float64(1), int64(6), object(5)
memory usage: 2.0+ MB


### Вывод

Перевели total_income (ежемесячный доход) из float64 в int64. Использовали метод astype(), потому что нам нужны целые (int), а to_numeric() переводит во float и нам в даном случае не подходит

Еще можно было бы сделать следующее:
- перевести days_employed (стаж работы) из float64 в int64;
- перевести gender из object в int64 - F будет 0, а M - 1.

Но поскольку эти столбцы нам не нужны в рамках даного проекта, не будем этим заниматься.

### Обработка дубликатов

In [5]:
print('Дубликатов:',data.duplicated().sum()) #ищем дубликаты
data = data.drop_duplicates().reset_index(drop = True) #удаляем дубликаты
print('\nРазмер таблицы: ',data.shape) #проверяем размер таблицы после удаления

Дубликатов: 54

Размер таблицы:  (21471, 12)


### Вывод

Удалили дубликаты, размер таблицы сократился с 21525 до 21471 записи.
Для поиска использовался метод duplicated() вместе с sum(), а для удаления - drop_duplicates() со сбросом индексов (а я других методов пока и не знаю).
В появлении дублей возможно виноват человеческий фактор, когда при занесении клиента вместо создания новой записи копировали уже существующую, но в дальнейшем по какой-то причине не вносили в нее изменения.

### Лемматизация

In [6]:
def LemmatizeIt(purpose):
    lemmas = m.lemmatize(purpose)
    if 'жилье' in lemmas or 'недвижимость' in lemmas:
        return 'недвижимость'
    if 'автомобиль' in lemmas:
        return 'автомобиль'
    if 'образование' in lemmas:
        return 'образование'
    return 'свадьба'

#смотрим, какие варианты вообще есть
print('Цели кредита\n',data['purpose'].unique()) 

#категоризируем цель кредита после лемматизации
data['purpose'] = data['purpose'].apply(LemmatizeIt)
print('\nЦели кредита после лемматизации и категоризации\n',data['purpose'].unique()) #проверили


Цели кредита
 ['покупка жилья' 'приобретение автомобиля' 'дополнительное образование'
 'сыграть свадьбу' 'операции с жильем' 'образование'
 'на проведение свадьбы' 'покупка жилья для семьи' 'покупка недвижимости'
 'покупка коммерческой недвижимости' 'покупка жилой недвижимости'
 'строительство собственной недвижимости' 'недвижимость'
 'строительство недвижимости' 'на покупку подержанного автомобиля'
 'на покупку своего автомобиля' 'операции с коммерческой недвижимостью'
 'строительство жилой недвижимости' 'жилье'
 'операции со своей недвижимостью' 'автомобили' 'заняться образованием'
 'сделка с подержанным автомобилем' 'получение образования' 'автомобиль'
 'свадьба' 'получение дополнительного образования' 'покупка своего жилья'
 'операции с недвижимостью' 'получение высшего образования'
 'свой автомобиль' 'сделка с автомобилем' 'профильное образование'
 'высшее образование' 'покупка жилья для сдачи' 'на покупку автомобиля'
 'ремонт жилью' 'заняться высшим образованием']

Цели кредита п

### Вывод

Лемматизировали цели кредита и заодно категоризировали их, сократив до 4 категорий.
Лемматизировали - значит брали поочередно в каждой табличной записи цель кредита и приводили каждое слово в ней к словарной форме (лемме). 

### Категоризация данных

In [7]:
def FamilyStatusCategorizer (family_status):
    family_status_single = ['Не женат / не замужем','в разводе','вдовец / вдова','гражданский брак']
    if family_status in family_status_single:
        return 'не женат / не замужем'
    return 'женат/замужем'

def FamilyStatusIdCategorizer (family_status_id):
    family_status_id_single = [4,3,2,1]
    if family_status_id in family_status_id_single:
        return 0
    return 1

def TotalIncomeGroup(total_income):
    if total_income < 100000:
        return 'менее 100 т.р.'
    if total_income < 200000:
        return 'от 100 до 200 т.р.'
    if total_income < 300000:
        return 'от 200 до 300 т.р.'
    return 'более 300 т.р.'

def ChildrenGroup(children):
    if children == 0:
        return 0
    return 1

#категоризируем семейный статус в две категории: женат/замужем и не женат/не замужем 
#все остальные нюансы банку думаю не интересны, поскольку юридически никак не оформлены
data['family_status'] = data['family_status'].apply(FamilyStatusCategorizer)

#также категоризируем id семеного статуса (1 - женат/замужем, 0 - не женат/не замужем)
data['family_status_id'] = data['family_status_id'].apply(FamilyStatusIdCategorizer)
print('Семейный статус\n',data['family_status'].value_counts()) #проверили
print('\nID cемейного статуса\n',data['family_status_id'].value_counts()) #проверили

#категоризируем уровень дохода в месяц (менее 100, от 100 до 200, от 200 до 300, более 300)
data['total_income_group'] = data['total_income'].apply(TotalIncomeGroup)
print('\nУровень дохода в месяц\n',data['total_income_group'].value_counts().sort_values()) #проверили

#категоризируем количество детей в наличие детей: есть(1)/нет(0), поскольку нам интересно наличие, а не количество
print('\nКоличество детей\n',data['children'].value_counts()) #смотрим, какие есть значения и сколько их
#ошибочные записи -1 (47) и 20 (76) - будем считать, что это 1 и 2 - их все равно мало и они нам погоду не сделают, 
#поскольку бездетных все равно в два раза больше
data['children_group'] = data['children'].apply(ChildrenGroup)
print('\nНаличие детей\n',data['children_group'].value_counts()) #проверили

Семейный статус
 женат/замужем            12344
не женат / не замужем     9127
Name: family_status, dtype: int64

ID cемейного статуса
 1    12344
0     9127
Name: family_status_id, dtype: int64

Уровень дохода в месяц
 более 300 т.р.         1483
от 200 до 300 т.р.     3584
менее 100 т.р.         4463
от 100 до 200 т.р.    11941
Name: total_income_group, dtype: int64

Количество детей
  0     14107
 1      4809
 2      2052
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

Наличие детей
 0    14107
1     7364
Name: children_group, dtype: int64


In [8]:
#выделим в отдельный словарь только те столбцы, которые нужны для анализа, и с ними уже и будем работать
analys_data = data[['children_group','family_status','total_income_group','purpose','debt']]
analys_data.head() #смотрим на нашу новую таблицу

children_group          family_status  total_income_group       purpose  \
0               1          женат/замужем  от 200 до 300 т.р.  недвижимость   
1               1          женат/замужем  от 100 до 200 т.р.    автомобиль   
2               0          женат/замужем  от 100 до 200 т.р.  недвижимость   
3               1          женат/замужем  от 200 до 300 т.р.   образование   
4               0  не женат / не замужем  от 100 до 200 т.р.       свадьба   

   debt  
0     0  
1     0  
2     0  
3     0  
4     0

### Вывод

Категоризировали семейный статус (family_status) и id семейного статуса (family_status_id) до 2 состояний: женаты(1)/не женаты(0), поскольку остальные категории не имеют для банка значения. 
Можно было бы вынести семейный статус в отдельный словарь, но у нас всего две категории - смысл? 

Категоризировали уровень дохода клиента (total_income)- объединили в несколько категорий с шагом 100 т.р. (если делать шаг меньше, в некоторых категориях будет мало значений - не из чего будет выводы делать

Категоризировали количество детей в наличие детей (есть -1/нет -0), поскольку вопрос стоит о зависимости между наличием детей и возвратом кредита, да и записей с количеством детей больше двух слишком мало.

Еще можно было бы категоризировать следующие стобцы:
- возраст клиента (dob_days) - перевести в возрастные группы
- трудовой стаж клиента (das_employed) - перевести в года и объединить в группы
- образование клиента (education) + вынести в отдельный словарь
- тип занятости клиента (income_type)

Но поскольку эти столбцы нам не нужны в рамках даного проекта, не будем этим заниматься.


### Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [9]:
#делаем сводную таблицу по числу клиентов
children_data_pivot_count = analys_data.pivot_table(index='children_group', values='debt', aggfunc='count') 
#делаем сводную таблицу по числу должников
children_data_pivot_sum = analys_data.pivot_table(index='children_group', values='debt', aggfunc='sum')
#объединяем
children_data_pivot = children_data_pivot_count.merge(children_data_pivot_sum, on='children_group', how='left')
#переобзываем столбцы
children_data_pivot.set_axis(['clients','debitors'],axis='columns',inplace=True)
#высчитываем доли должников
children_data_pivot['ratio,%'] = round((children_data_pivot['debitors'] / children_data_pivot['clients'])*100,1)
#сортируем по доле должников по возрастанию и смотрим на нашу таблицу
children_data_pivot.sort_values(by='ratio,%').head()


clients  debitors  ratio,%
children_group                            
0                 14107      1063      7.5
1                  7364       678      9.2

### Вывод

Люди без детей почти вдвое чаще людей с детьми берут кредиты, при этом особой зависимости между наличием детей и возвратом кредита в срок не наблюдается - должники есть и в той и в другой социальной группе, при этом разница в доли должников не более 1.7%. 

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [10]:
#делаем сводную таблицу по числу клиентов
family_data_pivot_count = analys_data.pivot_table(index='family_status', values='debt', aggfunc='count') 
#делаем сводную таблицу по числу должников
family_data_pivot_sum = analys_data.pivot_table(index='family_status', values='debt', aggfunc='sum')
#объединяем
family_data_pivot = family_data_pivot_count.merge(family_data_pivot_sum, on='family_status', how='left')
#переобзываем столбцы
family_data_pivot.set_axis(['clients','debitors'],axis='columns',inplace=True)
#высчитываем доли должников
family_data_pivot['ratio,%'] = round((family_data_pivot['debitors'] / family_data_pivot['clients'])*100,1)
#сортируем по доле должников по возрастанию и смотрим на нашу таблицу
family_data_pivot.sort_values(by='ratio,%').head()

clients  debitors  ratio,%
family_status                                    
женат/замужем            12344       931      7.5
не женат / не замужем     9127       810      8.9

### Вывод

Зависимости между семейным положением и возвратом кредита в срок также не наблюдается - должники есть и в той и в другой социальной группах, разница между долями должников не более 1.4%

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [11]:
#делаем сводную таблицу по числу клиентов
income_data_pivot_count = analys_data.pivot_table(index='total_income_group', values='debt', aggfunc='count') 
#делаем сводную таблицу по числу должников
income_data_pivot_sum = analys_data.pivot_table(index='total_income_group', values='debt', aggfunc='sum')
#объединяем
income_data_pivot = income_data_pivot_count.merge(income_data_pivot_sum, on='total_income_group', how='left')
#переобзываем столбцы
income_data_pivot.set_axis(['clients','debitors'],axis='columns',inplace=True)
#высчитываем доли должников
income_data_pivot['ratio,%'] = round((income_data_pivot['debitors'] / income_data_pivot['clients'])*100,1)
#сортируем по доле должников по возрастанию и смотрим на нашу таблицу
income_data_pivot.sort_values(by='ratio,%').head()

clients  debitors  ratio,%
total_income_group                            
от 200 до 300 т.р.     3584       252      7.0
более 300 т.р.         1483       106      7.1
менее 100 т.р.         4463       354      7.9
от 100 до 200 т.р.    11941      1029      8.6

### Вывод

Опять же, уровень дохода мало связан с возвратом кредита в срок - должники есть в каждой группе, разница между долями должников не более 1,6%

- Как разные цели кредита влияют на его возврат в срок?

In [12]:
#делаем сводную таблицу по числу клиентов
purpose_data_pivot_count = analys_data.pivot_table(index='purpose', values='debt', aggfunc='count') 
#делаем сводную таблицу по числу должников
purpose_data_pivot_sum = analys_data.pivot_table(index='purpose', values='debt', aggfunc='sum')
#объединяем
purpose_data_pivot = purpose_data_pivot_count.merge(purpose_data_pivot_sum, on='purpose', how='left')
#переобзываем столбцы
purpose_data_pivot.set_axis(['clients','debitors'],axis='columns',inplace=True)
#высчитываем доли должников
purpose_data_pivot['ratio,%'] = round((purpose_data_pivot['debitors'] / purpose_data_pivot['clients'])*100,1)
#сортируем по доле должников по возрастанию и смотрим на нашу таблицу
purpose_data_pivot.sort_values(by='ratio,%').head()

clients  debitors  ratio,%
purpose                                 
недвижимость    10814       782      7.2
свадьба          2335       186      8.0
образование      4014       370      9.2
автомобиль       4308       403      9.4

### Вывод

Чаще всего за кредитом обращаются при сделках с недвижимостью. При этом цель кредита никак не связана с вероятностью возникновения долга по кредиту - должники есть в каждой категории, разница между долями должников не более 2.2%. 

### Шаг 4. Общий вывод

По какому-то одному параметру (дети, семейное положение, уровень дохода, цель кредита) нельзя заранее рассчитать вероятность возникновения долга по кредиту. Надо суммировать эти параметры и уже из общей оценки делать выводы. Т.е. если человек не женат/не замужем, у него/нее есть ребенок/дети, ежемесячный доход между 100 и 200 т.р., а кредит он/она берет на автомобиль - по сумме этих параметров он/она будет в максимальной группе риска. 